File: src/main/scala/mmlTests/a201603a/javaTest/gpt35turbo/BagFactoryImpl.java
package mmlTests.a201603a.javaTest.gpt35turbo;

import mmlTests.a201603a.javaTest.Bag;
import mmlTests.a201603a.javaTest.BagFactory;
import java.util.*;
import java.util.function.*;

public class BagFactoryImpl implements BagFactory {
    public <X> Bag<X> empty() {
        return new BagImpl<X>(new HashMap<X,Integer>());
    }

    public <X> Bag<X> fromSet(Set<X> set) {
        Map<X, Integer> map = new HashMap<X, Integer>();
        for (X element : set) {
            map.put(element, 1);
        }
        return new BagImpl<X>(map);
    }

    public <X> Bag<X> fromList(List<X> list) {
        Map<X, Integer> map = new HashMap<X, Integer>();
        for (X element : list) {
            int count = map.getOrDefault(element, 0);
            map.put(element, count + 1);
        }
        return new BagImpl<X>(map);
    }

    public <X> Bag<X> bySupplier(Supplier<X> supplier, int nElements, ToIntFunction<X> copies) {
        Map<X, Integer> map = new HashMap<X, Integer>();
        for (int i = 0; i < nElements; i++) {
            X element = supplier.get();
            int count = copies.applyAsInt(element);
            map.put(element, count);
        }
        return new BagImpl<X>(map);
    }

    public <X> Bag<X> byIteration(X first, UnaryOperator<X> next, int nElements, ToIntFunction<X> copies) {
        Map<X, Integer> map = new HashMap<X, Integer>();
        X element = first;
        for (int i = 0; i < nElements; i++) {
            int count = copies.applyAsInt(element);
            map.put(element, count);
            element = next.apply(element);
        }
        return new BagImpl<X>(map);
    }
}

File: src/main/scala/llmTests/a201603a/scala/gpt4/BagFactoryImpl.scala
package llmTests.a201603a.scala.gpt4

import llmTests.a201603a.scala.Bag
import llmTests.a201603a.scala.BagFactory
import java.util
import java.util.function.{Supplier, ToIntFunction, UnaryOperator}
import scala.jdk.CollectionConverters._
import scala.collection.immutable.List

class BagFactoryImpl extends BagFactory{
  override def empty[X]: Bag[X] = new BagImpl[X](Nil)

  override def fromSet[X](set: util.Set[X]): Bag[X] = {
    new BagImpl[X](set.asScala.toList)
  }

  override def fromList[X](list: util.List[X]): Bag[X] = {
    new BagImpl[X](list.asScala.toList)
  }

  override def bySupplier[X](supplier: Supplier[X], nElements: Int, copies: ToIntFunction[X]): Bag[X] = {
    val list = List.fill(nElements)(supplier.get()).flatMap(x => List.fill(copies.applyAsInt(x))(x))
    new BagImpl[X](list)
  }

  override def byIteration[X](first: X, next: UnaryOperator[X], nElements: Int, copies: ToIntFunction[X]): Bag[X] = {
    var element = first
    var list = List.fill(copies.applyAsInt(element))(element)
    for (_ <- 1 until nElements) {
      element = next.apply(element)
      list = list ++ List.fill(copies.applyAsInt(element))(element)
    }
    new BagImpl[X](list)
  }

  private class BagImpl[X](val elems: List[X]) extends Bag[X]{
    override def numberOfCopies(x: X): Int = elems.count(_ == x)

    override def size: Int = elems.size

    override def toList: util.List[X] = elems.asJava
  }
}